### PR TITLE
[BUGFIX canary] Deprecate passing function as test argument to deprecate/warn/assert

### PR DESCRIPTION
**Pull request by e00dan**
This PR attempts to deprecate using function as test argument in: `Ember.deprecate`, `Ember.warn` and `Ember.assert`. It's first part of closing issue #11898. Tests for `ember-debug` pass, but I still have to fix other tests and code, which trigger deprecation with function as test argument. Seems like this behavior is used in many places in application lifecycle. It's my first attempt to commit actual code to Ember.js so I appreciate any help.

@mmun, @rwjblue

### CODE REVIEW

**Comment by Kuzirashi** [comment]
Tests pass. @mmun, @rwjblue

**Comment by Kuzirashi** [comment]
@rwjblue should I rebase or you aren't interested in this pr?

**Comment by stefanpenner** [comment]
i would be in favor, but will defer to @rwjblue

**Comment by Kuzirashi** [comment]
@rwjblue any feedback on this?

**Comment by mmun** [comment]
@Kuzirashi Sorry for no feedback. I'm definitely :+1: on this.

**Comment by Kuzirashi** [comment]
Thanks @mmun! I'm waiting for signal from @rwjblue and I'll eventually rebase and squash.

File: packages/ember-debug/lib/deprecate.js [hunk 1]
```diff
@@ -87,7 +87,7 @@ export let missingOptionsUntilDeprecation = 'When calling `Ember.deprecate` you
   @param {String} message A description of the deprecation.
   @param {Boolean|Function} test A boolean. If falsy, the deprecation
     will be displayed. If this is a function, it will be executed and its return
```
Comment (rwjblue) [inline packages/ember-debug/lib/deprecate.js]: Lets just remove the reference to using a function all together.

File: packages/ember-debug/lib/handlers.js [hunk 2]
```diff
@@ -1,9 +1,23 @@
 import isPlainFunction from 'ember-debug/is-plain-function';
+import deprecate from 'ember-debug/deprecate';
 
 export let HANDLERS = { };
+export let testAsFunctionDeprecation = 'Functional arguments are ambiguous with constructors. ' +
```
Comment (rwjblue) [inline packages/ember-debug/lib/handlers.js]: I think that we need to mention what they did that triggered the message. It makes this slightly annoying (due to passing through what was called), but I think it will make it much easier to track down.

One possible solution might be:

```javascript
export function generateTestAsFunctionDeprecation(source) {
  return `Calling \`${source}\` with a function argument is deprecated. Please ` +
    `use \`!!Constructor\` for constructors, or an `IIFE` to compute the test for deprecation. ` +
    `In a future version functions will be treated as truthy values instead of being executed.`
}
```

Then you would have to pass through `'Ember.' + type` in `invoke` function below (and thread that through to `normalizeTest`).

File: packages/ember-debug/lib/handlers.js [hunk 3]
```diff
@@ -1,9 +1,23 @@
...
+  'Please use !!Constructor for constructors, or an IIFE to compute the deprecation value. ' +
+  'In a future version of Ember functions will be treated as truthy values instead of being executed.';
 
 function normalizeTest(test) {
-  return isPlainFunction(test) ? test() : test;
+  if (isPlainFunction(test)) {
+    deprecate(
+      testAsFunctionDeprecation,
+      false,
+      { id: 'ember-debug.deprecate-test-as-function', until: '3.0.0' }
```
Comment (rwjblue) [inline packages/ember-debug/lib/handlers.js]: Lets use `until: '2.5.0'` for this.  I am uncertain if this actually a public API (IMHO the current behavior of invoking constructors feels like a bug), but if we get this deprecated in 2.2 we will have a couple versions for things to get fixed...

File: packages/ember-debug/lib/warn.js [hunk 4]
```diff
@@ -24,8 +24,9 @@ export let missingOptionsIdDeprecation = 'When calling `Ember.warn` you must pro
 
   @method warn
   @param {String} message A warning to display.
-  @param {Boolean} test An optional boolean. If falsy, the warning
-    will be displayed.
+  @param {Boolean|Function} test An optional boolean. If falsy, the warning
+    will be displayed. If this is a function, it will be executed and its return
```
Comment (rwjblue) [inline packages/ember-debug/lib/warn.js]: Lets not mention the `function` bit in the docs...

File: packages/ember-debug/tests/main_test.js [hunk 5]
```diff
@@ -105,14 +105,12 @@ QUnit.test('Ember.deprecate throws deprecation if second argument is falsy', fun
...
-QUnit.test('Ember.deprecate does not throw deprecation if second argument is a function and it returns true', function() {
+QUnit.test('Ember.deprecate throws deprecation if second argument is a function and it returns true', function() {
   expect(1);
 
-  Ember.deprecate('Deprecation is thrown', function() {
-    return true;
-  }, { id: 'test', until: 'forever' });
-
-  ok(true, 'deprecation was not thrown');
+  throws(() => {
```
Comment (rwjblue) [inline packages/ember-debug/tests/main_test.js]: need to use `expectDeprecation` here to allow tests to work properly when debugging methods are stripped (production builds).

File: packages/ember-debug/tests/main_test.js [hunk 6]
```diff
@@ -150,14 +148,10 @@ QUnit.test('Ember.assert throws if second argument is falsy', function() {
...
-QUnit.test('Ember.assert does not throw if second argument is a function and it returns true', function() {
+QUnit.test('Ember.assert throws if second argument is a function and it returns true', function() {
   expect(1);
 
-  Ember.assert('Assertion is thrown', function() {
-    return true;
-  });
-
-  ok(true, 'assertion was not thrown');
+  throws(() => Ember.assert('Assertion is thrown', () => true));
```
Comment (rwjblue) [inline packages/ember-debug/tests/main_test.js]: Should use `expectAssertion` here

File: packages/ember-htmlbars/lib/node-managers/view-node-manager.js [hunk 7]
```diff
@@ -27,13 +27,9 @@ function ViewNodeManager(component, scope, renderNode, block, expectElement) {
 export default ViewNodeManager;
 
 ViewNodeManager.create = function ViewNodeManager_create(renderNode, env, attrs, found, parentView, path, contentScope, contentTemplate) {
-  assert('HTMLBars error: Could not find component named "' + path + '" (no component or template with that name was found)', function() {
```
Comment (rwjblue) [inline packages/ember-htmlbars/lib/node-managers/view-node-manager.js]: Instead of rewriting to avoid the function, can you make the function an IIFE instead?  I personally find the function style of writing this much easier to understand.

File: packages/ember-htmlbars/tests/node-managers/view-node-manager-test.js [hunk 8]
```diff
@@ -0,0 +1,63 @@
+import ViewNodeManager from 'ember-htmlbars/node-managers/view-node-manager';
```
Comment (rwjblue) [inline packages/ember-htmlbars/tests/node-managers/view-node-manager-test.js]: THANK YOU FOR WRITING TESTS FOR THIS!!!!

File: packages/ember-views/lib/system/build-component-template.js [hunk 9]
```diff
@@ -294,8 +294,6 @@ function normalizeClasses(classes, output, streamBasePath) {
 }
 
 function validateTaglessComponent(component) {
-  assert('You cannot use `classNameBindings` on a tag-less component: ' + component.toString(), function() {
```
Comment (rwjblue) [inline packages/ember-views/lib/system/build-component-template.js]: Would you mind tweaking this into an IIFE to keep the function style?

File: packages/ember-htmlbars/tests/node-managers/view-node-manager-test.js (hunk 8 above)
Comment (mmun) [inline packages/ember-htmlbars/tests/node-managers/view-node-manager-test.js]: :P *Pssst. There's component-node-manager too.*

I was thinking this should be pulled out to another PR so it doesn't interfere with the [BUGFIX release].

**Comment by rwjblue** [comment]
@Kuzirashi - This looks very good. I have left a few pretty minor comments. When you have time can you fixup the comments, squash, and rebase?

**Comment by Kuzirashi** [comment]
I'll find the time. I'll mention you when it's all ready. Thanks for review. ;)

File: packages/ember-htmlbars/tests/node-managers/view-node-manager-test.js (hunk 8 above)
Comment (rwjblue) [inline packages/ember-htmlbars/tests/node-managers/view-node-manager-test.js]: These tests are to confirm the [changes in the assertion](https://github.com/emberjs/ember.js/pull/12370/files#r40686699) work properly (that is why there isn't a corresponding change to `component-node-manager.js`).

Comment (Kuzirashi) [inline packages/ember-htmlbars/tests/node-managers/view-node-manager-test.js]: Yeah, I needed it to help me confirm I kept the same logic.

**Comment by rwjblue** [comment]
I agree with @mmun though, this deprecation should probably not be  `[BUGFIX release]`.  In order to allow addons to deal with the new deprecation in an orderly way, I'd like to land this on master before branching into 2.2 beta (which should be in the next few days).  That gives folks the full beta cycle (~6weeks) to get things updated in addons so this doesn't impact end user apps much.

tldr; can you update to use `[BUGFIX canary]` instead?

**Comment by Kuzirashi** [comment]
Sure it'll be changed to `[BUGFIX canary]`.

File: packages/ember-htmlbars/tests/node-managers/view-node-manager-test.js (hunk 8 above)
Comment (mmun) [inline packages/ember-htmlbars/tests/node-managers/view-node-manager-test.js]: Ok :+1:

File: packages/ember-debug/tests/main_test.js (hunk 6 above)
Comment (Kuzirashi) [inline packages/ember-debug/tests/main_test.js]: It shouldn't expectAssertion. It should expectDeprecation, because Ember.assert will throw only if it receives FAILING test. Am I right? @rwjblue

File: packages/ember-debug/tests/main_test.js (hunk 5 above)
Comment (Kuzirashi) [inline packages/ember-debug/tests/main_test.js]: It doesn't work when using `expectDeprecation` seems like it doesn't even call invoke.

File: packages/ember-debug/tests/main_test.js (hunk 6 above)
Comment (rwjblue) [inline packages/ember-debug/tests/main_test.js]: Yes, that's right (I read too quickly).

File: packages/ember-debug/tests/main_test.js (hunk 5 above)
Comment (Kuzirashi) [inline packages/ember-debug/tests/main_test.js]: It works fine though when moving Ember.deprecate out of expectDeprecation callback or using throws. I think it won't be possible with using expectDeprecation. I can't debug it, I don't know how that helper works. Seems like triggering deprecation from inside of Ember.deprecate is too much for it. In other places in tests in this file throws is used. Can I stick to using throws, or should I go with registerHandler approach? @rwjblue

Comment (rwjblue) [inline packages/ember-debug/tests/main_test.js]: Use throws for now (though I think expectDeprecation should work), we can circle back if we have issues with CI on production builds.

**Comment by Kuzirashi** [comment]
Rebased, squased, tests pass. Except tests on sauce which randomly pass or fail from my experience. @rwjblue

**Comment by rwjblue** [comment]
Awesome, thanks for your hard work on this @Kuzirashi!

**Comment by rwjblue** [comment]
@Kuzirashi - Would you mind submitting a PR to https://github.com/emberjs/website/blob/master/source/deprecations/v2.x.html.md adding a section on this for 2.2?

**Comment by Kuzirashi** [comment]
Sure :)

**Comment by Kuzirashi** [comment]
PR in website repo created. @rwjblue
